### PR TITLE
Handle e-mails with unsupported encoding better

### PR DESCRIPTION
**Pull request by philipwhiuk**
From this stack-trace:

```
K9-Mail version: 5.115
Device make: samsung
Device model: GT-N5100
Android version: 4.4.4

com.fsck.k9.mailstore.LockableDatabase$WrappedException: com.fsck.k9.mail.MessagingException: Error appending message: Tworzenie stron WWW, urządzenia mobilne - Pozycjonowanie
 at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1090)
 at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1078)
 at com.fsck.k9.mailstore.LockableDatabase.execute(LockableDatabase.java:275)
 at com.fsck.k9.mailstore.LocalFolder.storeSmallMessage(LocalFolder.java:1078)
 at com.fsck.k9.controller.MessagingController$12.messageFinished(MessagingController.java:1406)
 at com.fsck.k9.mail.store.imap.ImapFolder.fetch(ImapFolder.java:770)
 at com.fsck.k9.controller.MessagingController.downloadSmallMessages(MessagingController.java:1393)
 at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1196)
 at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:953)
 at com.fsck.k9.controller.MessagingController$36.run(MessagingController.java:4010)
 at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
 at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
 at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
 at java.lang.Thread.run(Thread.java:841)
Caused by: com.fsck.k9.mail.MessagingException: Error appending message: Tworzenie stron WWW, urządzenia mobilne - Pozycjonowanie
 at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1326)
 at com.fsck.k9.mailstore.LocalFolder$14.doDbWork(LocalFolder.java:1187)
 at com.fsck.k9.mailstore.LocalFolder$14.doDbWork(LocalFolder.java:1182)
 at com.fsck.k9.mailstore.LockableDatabase.execute(LockableDatabase.java:275)
 at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:1182)
 at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:1109)
 at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1082)
 ... 13 more
Caused by: java.lang.UnsupportedOperationException: Encoding for RawDataBody not supported: utf-8
 at com.fsck.k9.mail.internet.MimeUtility.decodeBody(MimeUtility.java:1047)
 at com.fsck.k9.mail.internet.MessageExtractor.getTextFromTextPart(MessageExtractor.java:105)
 at com.fsck.k9.mail.internet.MessageExtractor.getTextFromPart(MessageExtractor.java:55)
 at com.fsck.k9.crypto.MessageDecryptVerifier.isPartPgpInlineEncrypted(MessageDecryptVerifier.java:207)
 at com.fsck.k9.message.extractors.EncryptionDetector.containsInlinePgpEncryptedText(EncryptionDetector.java:34)
 at com.fsck.k9.message.extractors.EncryptionDetector.isEncrypted(EncryptionDetector.java:25)
 at com.fsck.k9.message.extractors.MessagePreviewCreator.createPreview(MessagePreviewCreator.java:31)
 at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1251)
 ... 19 more
```

Now, "utf-8" is a charset and not an encoding. So there is no issue in my eyes that K-9 does not display a message.

However we should not break sync either.

Ideally I'd like to add an `ErrorPart` in the future for issues that arise when parsing an individual email. This would give the user an option to 'Re-try download' for I/O errors (moving it from the context menu which is unintuitive) and explain the issue for others..

In the shorter term however it's clearly rubbish that a single badly encoded (and typically (because they don't care) spam) email should break sync.

Hence I propose the commit attached. This will produce a blank e-mail. Not ideal but allows the user to get the rest of their messages.

### CODE REVIEW

**Comment by Valodim** [comment]
There is already an UnsupportedEncodingException in the standard lib, which used for charsets. We should definitely use a different name here :)

**Comment by philipwhiuk** [comment]
If it weren't for the fact this was an uncaught exception it wouldn't have broken sync. I think we do handle them. Hence why we get blank emails / emails with only structure - we've synced an email but couldn't decode it (previously this occurred a fair bit because of that I/O error you fixed - #1243 - now it should hopefully just be the odd issue like this where the email is borked/unsupported).

If we add the ErrorPart then we'll fix that issue because we'll show something useful to the user (and be able to kill off more of what goes in K9mail-errors).

**Comment by Valodim** [comment]
Yeah, seems good to me.

For error handling: I think we can just fail and display an error for thoroughly broken emails such as this. As mitigation path, we should focus on offering diagnostics like "display source", but generally there's not much we can do

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/UnsupportedContentTransferEncodingException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.fsck.k9.mail.internet;
+
+public class UnsupportedContentTransferEncodingException extends Exception {
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/UnsupportedContentTransferEncodingException.java]: Why not make this a subclass of `MessagingException` so we don't need to change all `throws` clause?

File: k9mail-library/src/test/java/com/fsck/k9/mail/internet/MessageExtractorTest.java [hunk 2]
```diff
@@ -72,13 +72,13 @@ public void getTextFromPart_withExceptionThrownGettingInputStream_shouldReturnNu
...
-    @Test(expected = UnsupportedOperationException.class)
-    public void getTextFromPart_withUnknownEncoding_shouldThrowRuntimeException() throws Exception {
+    @Test
+    public void getTextFromPart_withUnknownEncoding_shouldReturnNull() throws Exception {
         part.setHeader(MimeHeader.HEADER_CONTENT_TYPE, "text/plain");
         BinaryMemoryBody body = new BinaryMemoryBody("Sample text body".getBytes(), "unknown encoding");
         part.setBody(body);
 
-        MessageExtractor.getTextFromPart(part);
+        assertNull(MessageExtractor.getTextFromPart(part));
```
Comment (cketti) [inline k9mail-library/src/test/java/com/fsck/k9/mail/internet/MessageExtractorTest.java]: Arrange, act, assert.

```java
…

String textFromPart = MessageExtractor.getTextFromPart(part) ;

assertNull(textFromPart);
```

File: k9mail-library/src/test/java/com/fsck/k9/mail/internet/MimeMessageParseTest.java [hunk 3]
```diff
@@ -128,6 +128,23 @@ public void testMultipartSingleLayerNoRecurse() throws Exception {
...
+                        "To: <eva@example.org>\r\n" +
+                        "Subject: Testmail\r\n" +
+                        "MIME-Version: 1.0\r\n" +
+                        "Content-type: text/plain\r\n" +
+                        "Content-Transfer-Encoding: utf-8\r\n" +
+                        "\r\n" +
+                        "dGhpcyBpcyBzb21lIG1vcmUgdGVzdCB0ZXh0Lg==\r\n"));
+
+        checkAddresses(msg.getFrom(), "adam@example.org");
+        checkAddresses(msg.getRecipients(RecipientType.TO), "eva@example.org");
```
Comment (cketti) [inline k9mail-library/src/test/java/com/fsck/k9/mail/internet/MimeMessageParseTest.java]: These two checks seem unrelated to the actual check you want to perform.

**Comment by philipwhiuk** [comment]
Thanks - updated based on feedback in latest commit :)
